modules/dr-restoring-cluster-state.adoc

```diff
@@ -304,3 +304,15 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-2-restoring-cluster-state_storage_{context}"]
+= Issues and workarounds in restoring persistent storage state
+
+If your {product-title} uses persistent storage of any form, typically a state of the world is stored outside etcd. It could be a ElasticSearch
+cluster running in a Pod or a database running in a statefulset. When {product-title} is restored from etcd backup, we are also restoring status of those workloads in Openshift. But depending on how old the etcd snapshot is - that status may be invalid or outdated.
+
+Some examples are:
+
+1. Mysql database was running in a Pod backed up by a PV object. But after the snapshot was taken - the Mysql Pod and corresponding PV were deleted and hence underlying volume on storage provider would be deleted. Restoring Openshift from snapshot is not going to bring back the volume on storage provider and hence is not going to result in running Mysql Pod(even though Pod will try to start but will keep failing) and a manual action must be taken to restore mysql Pod. In this case manual action could be restoring the volume on storage provider via volume snapshots and then editing PV object to point to new volume.
+
+2. You have a pod P1 running using volume A which is attached to node X. If the snapshot is taken and another pod uses the same volume on node Y and a etcd restore is performed then pod P1 may not able to start correctly because volume is still attached to node Y but {product-title} does know about it and hence is not going to detach it. In this case volume must be manually detached from node Y, so as volume can attach on node X and pod P1 can start.
```

Comment:
Wont the storage provider eventually reconcile?  Give an example of a storage provider this won’t work on?

Reply:
It won't work for CSI(all CSI volumes) or GCE-PD intree driver for example. For intree driver's volume's attachment status is recorded in `node.Status` and similarly for CSI volumes in separate `VolumeAttachment` object. If a etcd restore causes `VolumeAttachment` to be removed or `node.Status` to be wiped - then attach/detach controller will not try and detach the volumes.   We have partially fixed this for some intree drivers by implementing a dangling volume mechanism but it isn't completely solved for CSI volumes - https://github.com/kubernetes/kubernetes/issues/80488 and some intree drivers (like GCE-PD) don't implement dangling volume mechanism at all. 

There is a opposite of this - which is, if when the snapshot was taken volume was attached to node X but since then pod got deleted and volume was detached but then snapshot is restored and pod starts on node X but since attach-detach controller thinks volume is *actually* attached already (because `node.Status` tells it so), it won't try to attach the volume again. This is also partially solved for some volume types but for CSI volumes `LIST_VOLUME` is an optional capability.

Reply:
Please note even though - we *now* have a dangling volume mechanism for CSI, if corresponding VolumeAttachment objects does not exist after restore from snapshot, volume will not be detached to correct the attachment issue.